Classes/Common/Indexer.php

```diff
@@ -336,6 +336,7 @@ protected static function processLogical(Document $document, array $logicalUnit)
             $solrDoc->setField('toplevel', $logicalUnit['id'] == $doc->toplevelId ? true : false);
             $solrDoc->setField('title', $metadata['title'][0], self::$fields['fieldboost']['title']);
             $solrDoc->setField('volume', $metadata['volume'][0], self::$fields['fieldboost']['volume']);
+            $solrDoc->setField('date', $metadata['date']);
```

Comment:
We need to make sure that `$metadata['date']` is always set when using it this way. Most documents will have a date, but there certainly are cases, where no date is given and therefore `$metadata['date']` is not set while retrieving the document's metadata. Please at least set `$metadata['date']` to an empty string or an appropriate default value in `MetsDocument->getMetadata()` and `IiifManifest->getMetadata()` to avoid this.

Furthermore, we have to validate the retrieved value and make sure it is properly formatted (i. e. according to XML's `dateTime`).

Reply:
The indexer now checks the date format. Dates in the format `YYYY`, `YYYY-MM`, `YYYY-MM-DD` will be kept as they are - especially `YYYY` and `YYYY-MM` to benefit from how the `dateRangeField` works with them. `YYYYMMDD` will be transformed to `YYYY-MM-DD`, to be complient to the necessary formating for this field type. Any `datetime` will be changed to the complient `YYYY-MM-DDThh:mm:ssZ` if necessary.  Empty and invalid dates will be skipped. Below a test script:

```
<?php
$dates = [
  "1987",
  "198710",
  "871002",
  "19871002",
  "19871002T19",
  "19871002T19Z",
  "19871002T1902",
  "19871002T1902Z",
  "19871002T190215",
  "19871002T190215Z",
  "19871002T190215.222",
  "19871002T190215.222Z",
  "1987",
  "1987-10",
  "87-10-02",
  "1987-10-02",
  "1987-10-02T19",
  "1987-10-02T19Z",
  "1987-10-02T19:02",
  "1987-10-02T19:02Z",
  "1987-10-02T19:02:15",
  "1987-10-02T19:02:15Z",
  "1987-10-02T19:02:15.222",
  "1987-10-02T19:02:15.222Z",
  
];

$isoDate = date('Y-m-d\TH:i:s\Z', strtotime('1987-10-02T19:02:15.222Z'));

echo $isoDate . "\n";
foreach ($dates as $date) {
  if (strtotime($date)) {
    echo $date . " is a valid date : " . date('c', strtotime($date));
  } else {
    echo $date . " is not a valid date!";
  }
  echo "\n";
}

?>
```
```
1987-10-02T19:02:15Z
1987 is a valid date : 1987-12-09T09:28:28+00:00
198710 is not a valid date!
871002 is not a valid date!
19871002 is a valid date : 1987-10-02T00:00:00+00:00
19871002T19 is not a valid date!
19871002T19Z is not a valid date!
19871002T1902 is a valid date : 1987-10-02T19:02:00+00:00
19871002T1902Z is a valid date : 1987-10-02T19:02:00+00:00
19871002T190215 is a valid date : 1987-10-02T19:02:15+00:00
19871002T190215Z is a valid date : 1987-10-02T19:02:15+00:00
19871002T190215.222 is not a valid date!
19871002T190215.222Z is not a valid date!
1987 is a valid date : 1987-12-09T09:28:28+00:00
1987-10 is a valid date : 1987-10-01T00:00:00+00:00
87-10-02 is a valid date : 1987-10-02T00:00:00+00:00
1987-10-02 is a valid date : 1987-10-02T00:00:00+00:00
1987-10-02T19 is not a valid date!
1987-10-02T19Z is not a valid date!
1987-10-02T19:02 is a valid date : 1987-10-02T19:02:00+00:00
1987-10-02T19:02Z is a valid date : 1987-10-02T19:02:00+00:00
1987-10-02T19:02:15 is a valid date : 1987-10-02T19:02:15+00:00
1987-10-02T19:02:15Z is a valid date : 1987-10-02T19:02:15+00:00
1987-10-02T19:02:15.222 is a valid date : 1987-10-02T19:02:15+00:00
1987-10-02T19:02:15.222Z is a valid date : 1987-10-02T19:02:15+00:00
```